sshd-core/src/main/java/org/apache/sshd/common/channel/BufferedIoOutputStream.java

```diff
@@ -52,60 +78,164 @@ public Object getId() {
+        if (pendingError != null) {
+            log.error("startWriting({})[{}] propagate to {} write requests pending error={}[{}]",
+                    getId(), out, writes.size(), getClass().getSimpleName(), pendingError.getMessage());
+
+            IoWriteFutureImpl currentFuture = currentWrite.getAndSet(null);
+            for (IoWriteFutureImpl pendingWrite : writes) {
+                // Checking reference by design
+                if (GenericUtils.isSameReference(pendingWrite, currentFuture)) {
+                    continue;   // will be taken care of when its listener is eventually called
+                }
+
+                future.setValue(pendingError);
+            }
+
+            writes.clear();
+            return;
+        }
+
+        // Cannot honor this request yet since other pending one incomplete
         if (!currentWrite.compareAndSet(null, future)) {
             return;
         }
 
-        out.writeBuffer(future.getBuffer()).addListener(
-                new SshFutureListener<IoWriteFuture>() {
-                    @Override
-                    public void operationComplete(IoWriteFuture f) {
-                        if (f.isWritten()) {
-                            future.setValue(Boolean.TRUE);
-                        } else {
-                            future.setValue(f.getException());
-                        }
-                        finishWrite(future);
-                    }
-                });
+        Buffer buffer = future.getBuffer();
+        int bufferSize = buffer.available();
+        out.writeBuffer(buffer).addListener(new SshFutureListener<IoWriteFuture>() {
+            @Override
+            public void operationComplete(IoWriteFuture f) {
+                if (f.isWritten()) {
+                    future.setValue(Boolean.TRUE);
+                } else {
+                    future.setValue(f.getException());
+                }
+                finishWrite(future, bufferSize);
+            }
+        });
     }
 
-    protected void finishWrite(IoWriteFutureImpl future) {
+    protected void finishWrite(IoWriteFutureImpl future, int bufferSize) {
+        /*
+         * Update the pending bytes count only if successfully written,
+         * otherwise signal an error
+         */
+        if (future.isWritten()) {
+            long writtenSize = writtenBytesCount.addAndGet(bufferSize);
+
+            int stillPending;
+            synchronized (pendingBytesCount) {
+                stillPending = pendingBytesCount.addAndGet(0 - bufferSize);
+                pendingBytesCount.notifyAll();
+            }
+
+            if (stillPending < 0) {
+                log.error("finishWrite({})[{}] - pending byte counts underflow ({}) after {} bytes", getId(), out, stillPending,
+                        writtenSize);
+                pendingException.compareAndSet(null,
+                        new SshChannelBufferedOutputException(channelId, "Pending byte counts underflow"));
+            }
```

Comment:
I'll review this code...

Reply:
Yes, I intended this - I am fine with a single pending write giving this mechanism "the slip". My goal was to avoid an OOM by having an unlimited accumulation of pending write requests due to fact that the peer is not consuming the sent data. Please note that the pending exception is "sticky" - i.e., the _next_ write attempt will fail. This also means that if the write request the "got away" was the last one by change and it was consumed by the peer there will be no exception thrown - which is also fine by me since as mentioned the goal is not to enforce a strict limit on the pending bytes size but rather on the accumulation of the pending write requests. I will document this in the code to make it clear.

Reply:
Adnedum: the main goal of the code is

>> to avoid an OOM by having an unlimited accumulation of pending write requests due to fact that the peer is not consuming the sent data

so we could have counted pending requests rather than bytes. However, we also want to avoid having a large amount of data pending consumption by the peer as well. This code strikes such a balance by allowing a _single_ pending request to exceed the limit, but at the same time prevents too many bytes from pending by having a bunch of pending requests that while below the imposed number limit may cumulatively represent a lot of pending bytes.